### ISSUE TITLE
Daylight Savings causing grouping issues in time_summarize, time_collapse, and time_group

### ISSUE THREAD

**Issue by tbradley1013** [business-science/tibbletime#31]
**tl:dr** Daylight Savings is causing `create_series` to select the wrong breakpoints for `1~d` period to be used for grouping

<hr>

I have been working with `time_summarise` recently for sub-daily data (2-minute interval) to try and take daily averages. It seems that `time_summarize` is taking the values from the first hour (12:00-12:59 AM) of the following day and grouping it with the previous day. It consequently is then ignored from the day it should be with. This behavior is, however, not consistent. With some days correctly grouping the day with the last value being in the last hour of that day (11:00-11:59 PM). I have put together a small example to help illustrate the problem using hourly data and your `create_series` function.

    # first create the hourly series and add a random variable to summarize
    x <- create_series(2016-03-12 ~ 2016-03-15, period = 1~h) %>%
      mutate(random_val = 1:nrow(x),
         
             # I do this so that the break is not right at midnight
             # I would expect that it should still classify 12:00 AM as
             # the following day, but I wanted to show that that is not 
             # the only occurance
             date = date + minutes(10))

Now before diving in too deep, let me show you what I mean. 

    x_sum <- x %>%
      time_summarise(1~d, 
                     mean = mean(random_val),
                     count = n())

    x_sum

    # A time tibble: 4 x 3
    # Index: date
                     date  mean count
    *              <dttm> <dbl> <int>
    1 2016-03-12 23:10:00  12.5    24
    2 2016-03-14 00:10:00  36.5    24
    3 2016-03-15 00:10:00  60.5    24
    4 2016-03-15 23:10:00  84.0    23

As you can see, for `2016-03-12` it grouped the values correctly, with 24 values and the final time being `2016-03-12 23:10:00` (the last hourly value for that day). But that is where things get weird. For the next day (`2016-03-13`), it groups the values but includes `2016-03-14 00:10:00` in the group, which results in the date displayed to be both misleading and the subsequent summaries to be incorrect. Interestingly, `2016-03-13 00:02:00` was daylight savings time. So there are only 23 values for that day in the created dataset.
    
    x %>%
      time_filter(~2016-03-13) %>%
      ungroup() %>%
      summarize(count = n(),
                first = first(date),
                last = last(date))

    # A tibble: 1 x 3
      count               first                last
      <int>              <dttm>              <dttm>
    1    23 2016-03-13 00:10:00 2016-03-13 23:10:00

To verify this, we can look at the number of values in the tibble for just `2016-03-13` and we can indeed see that there are 23 and that the first and last value of that day are `2016-03-13 00:10:00` and the last time is `2016-03-13 23:10:00`, as would be expected. This behavior is carried on until the final day in this example, with `2016-03-15` having only 23 values, since its midnight value is placed in another group. 

To compare what it looks like using pure `dplyr`, and to confirm (at least I think this confirms it) that this is not an issue with my computers local time zones or anything like that:

    x %>%
      mutate(date_mut = date(date)) %>%
      group_by(date_mut) %>%
      summarize(last = last(date),
                mean = mean(random_val),
                count = n())

    # A tibble: 4 x 4
        date_mut                last  mean count
          <date>              <dttm> <dbl> <int>
    1 2016-03-12 2016-03-12 23:10:00  12.5    24
    2 2016-03-13 2016-03-13 23:10:00  36.0    23
    3 2016-03-14 2016-03-14 23:10:00  59.5    24
    4 2016-03-15 2016-03-15 23:10:00  83.5    24

So it looks like the pure `dplyr` method worked judging solely by the `count` and the `last` columns. Quick mean calculations, not included here, also show the `mean` column is correct. 

Now, I have tried to dive into your code to figure out where the problem could be occuring, to try and save you guys some time. It looks like [`time_summarise`](https://github.com/business-science/tibbletime/blob/master/R/time_summarise.R#L72) is a `time_collapse`, `group_by`, `summarize` combo. So I backed up and just ran `time_collapse` to see if the issue was seen there.
 
    x_col <- x %>%
      time_collapse(1~d, as_sep_col = TRUE) 

    x_col[c(23, 24, 25, 46, 47, 48),]

    # A time tibble: 6 x 3
    # Index: date
                     date               .date random_val
    *              <dttm>              <dttm>      <int>
    1 2016-03-12 23:10:00 2016-03-12 22:10:00         23
    2 2016-03-12 23:10:00 2016-03-12 23:10:00         24
    3 2016-03-14 00:10:00 2016-03-13 00:10:00         25
    4 2016-03-14 00:10:00 2016-03-13 22:10:00         46
    5 2016-03-14 00:10:00 2016-03-13 23:10:00         47
    6 2016-03-14 00:10:00 2016-03-14 00:10:00         48

As we can see, the problem does occur within `time_collapse`, and is just carried forward into `time_summarize`. Looking through the [`time_collapse`](https://github.com/business-science/tibbletime/blob/master/R/time_collapse.R#L90) code, it looks like the issue is with the `time_group` function. 

    x_tg <- x %>% 
      mutate(.time_group = time_group(date, 1~d))

    x_tg[c(23, 24, 25, 46, 47, 48),]

    # A time tibble: 6 x 3
    # Index: date
                     date random_val .time_group
    *              <dttm>      <int>       <dbl>
    1 2016-03-12 22:10:00         23           1
    2 2016-03-12 23:10:00         24           1
    3 2016-03-13 00:10:00         25           2
    4 2016-03-13 22:10:00         46           2
    5 2016-03-13 23:10:00         47           2
    6 2016-03-14 00:10:00         48           2

As we can see, `time_group` is incorrectly grouping the first hour of `2016-03-14` with `2016-03-13`. Digging into the `time_group` code is looks like the issue is at the [`create_series`](https://github.com/business-science/tibbletime/blob/master/R/time_group.R#L115) function call. Running through the function step by step on my machine with `x$date` as my index and `1~d` as my period, `endpoint_dates` is :

    [1] "2016-03-12 00:00:00 EST" "2016-03-13 00:00:00 EST" "2016-03-14 01:00:00 EDT" "2016-03-15 01:00:00 EDT"
    [5] "2016-03-16 01:00:00 EDT"
    
So it looks like for some reason this is reading 1 AM as the start of each day for `2016-03-14` through `2016-03-16`. Doing this for an extended perioed shows that the behavior persists until the end of the range since there is not data for the following day. So it seems that daylight savings is throwing off this calculation which is then affecting every subsequent calculation.

I hope this helps you nail down where the problem is. I will look at `create_series` myself to see if I come up with any creative solutions. I believe my coworker ran into an issue with monthly summaries as well, so I will look into that and let you know if I am able to reporduce it. 

Let me know if you need anything else from me, I am happy to help in any way I can.

**Comment by tbradley1013** [comment]
I am sure you will probably want to address this in `create_series` itself, but one possible solution in `time_group` would be to use `floor_date(endpoint_dates, unit = period_list$period)`, but I am not sure if this would cause issues with other cases or not

**Comment by DavisVaughan** [comment]
Thanks for such a great full example, and for digging into it yourself.

This is at least partially caused by `seq.POSIXt()` inside `create_series()`. 

When I run this on the east coast, I get 2 different time zones in my date column, EST and EDT. It looks like you do too. This is due to using `Sys.Timezone()` to autodetect the user's time zone under the hood. 

I'm learning that it chooses `America/New_York`, which switches between `EST` and `EDT` at the daylight savings time change. Because of this, when creating the endpoints, once daylight savings hits, the next endpoint went from `"2016-03-14 00:00:00 EST"` -> `"2016-03-14 01:00:00 EDT"`.

The quick and dirty way to instantly deal with this is to specify a timezone:

```r
x <- create_series(2016-03-12 ~ 2016-03-15, period = 1~h, tz = "EST")
```

This would include 96 data points instead of 95, because the daylight savings switch isn't performed. Things work as you'd expect, but I think the counts are "wrong" depending on it you care about daylight savings:

```r
# A time tibble: 4 x 3
# Index: date
                 date  mean count
*              <dttm> <dbl> <int>
1 2016-03-12 23:10:00  12.5    24
2 2016-03-13 23:10:00  36.5    24
3 2016-03-14 23:10:00  60.5    24
4 2016-03-15 23:10:00  84.5    24
```

I'm thinking that maybe I should just set the default time zone for `create_series()` to `EST` if the user didn't select one. That way they don't run into these problems. 

In the meantime, I'll think about how to handle the `America/New_York` case where daylight savings is used.

**Comment by tbradley1013** [comment]
That makes sense, I didn't realize that it switches time zones during daylight savings. As for just specifying `tz` in my `create_series()`, that may fix the problem for the dummy dataset I made. But I am not sure if that will fix it when I am using my actual data. I think the big thing with regards to `time_summarize` and `time_collapse` would be what the `tz` argument in `create_series()` inside of `time_group()`  is set to something other than what the default is for `create_series`, unless you choose to not specify a `tz` inside that particular function call. 

Looking at the `attr(..., "tz")` for my actual data that I used when I had this issue the tz was `""`.  Switching the `tz` to "UTC" and rerunning, I do not get the issue. So I think you are correct that it is because my default `tz` in `America/New York`, which is then used in the `create_series()` inside `time_group()`.

**Comment by DavisVaughan** [comment]
Additionally, when you run `as_tbl_time()` it sets the "time_zone" attribute to `America/New_York` if you don't set anything. That's why even though your data had a `""` tz, you still had the problem

**Comment by DavisVaughan** [comment]
I've fixed in in a dev branch

```r
devtools::install_github("business-science/tibbletime", ref = "feature-rcpp-is_ordered")
```

**Comment by tbradley1013** [comment]
is there supposed to be a way to specify `tz` inside of `as_tbl_time()`? Looking at the function, it looks like it just uses either the timezone of the index column or the time zone of the system, but I am not seeing a way to specify it without manually setting it prior. 

If not, it's not a big deal, your comment just made it seem like there might be.

**Comment by DavisVaughan** [comment]
Nope, it does exactly what you said. The only time a time zone is specified is on creation of a new series by `create_series`. The time zone attr is set in `as_tbl_time()` (even on Date objects) because it gets passed through to `create_series` and something needs to be there even if it doesnt affect the calculation.

**Comment by tbradley1013** [comment]
Awesome, thanks for the clarifications and looking into this so quickly!
